### PR TITLE
Several Changes

### PR DESCRIPTION
**Pull request by Zhehui-Huang**
1. Modify some parameters of attention encoder, make neighbor obs and neighbor encoder orthogonal
2. Add distance to walls and floors to the observation space of drone.
3. Introduce smooth collision function given **distance** among drones
4. Introduce smooth collision function given **velocity** among drones
5. Minor modifications of neighbor obs type

### CODE REVIEW

File: envs/quadrotors/quad_utils.py [hunk 1]
```diff
@@ -78,7 +78,11 @@ def make_quadrotor_env_multi(cfg, **kwargs):
...
         quads_vel_reward_out_range=cfg.quads_vel_reward_out_range, quads_obstacle_mode=cfg.quads_obstacle_mode,
         quads_view_mode=cfg.quads_view_mode, quads_obstacle_num=cfg.quads_obstacle_num, quads_obstacle_type=cfg.quads_obstacle_type, quads_obstacle_size=cfg.quads_obstacle_size,
-        adaptive_env=cfg.quads_adaptive_env, obstacle_traj=cfg.quads_obstacle_traj, local_obs=cfg.quads_local_obs
+        adaptive_env=cfg.quads_adaptive_env, obstacle_traj=cfg.quads_obstacle_traj, local_obs=cfg.quads_local_obs, obs_repr=cfg.quads_obs_repr,
+        collision_hitbox_radius=cfg.quads_collision_hitbox_radius, collision_falloff_radius=cfg.quads_collision_falloff_radius,
+        collision_smooth_max_penalty=cfg.quads_collision_smooth_max_penalty, collision_vel_penalty_mode=cfg.quads_collision_vel_penalty_mode,
+        collision_smooth_vel_coeff=cfg.quads_collision_smooth_vel_coeff, collision_vel_penalty_radius=cfg.quads_collision_vel_penalty_radius,
+        collision_smooth_vel_max_penalty=cfg.quads_collision_smooth_vel_max_penalty
     )
```
Comment (alex-petrenko) [inline envs/quadrotors/quad_utils.py:86]: This is a bit crazy. Not sure we needed configurable parameters for all of these... but okay

File: runner/runs/quad_multi_mixed_vel_col_func.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    ('quads_collision_vel_penalty_mode', ['none', 'linear', 'quadratic']),
+    ('quads_collision_vel_penalty_radius', [4.0, 6.0]),
+    ('quads_collision_smooth_vel_coeff', [0.5, 2]),
+])
+
+_experiment = Experiment(
+    'grid_search_vel_collision_func-8_mixed',
+    'python -m run_algorithm --env=quadrotor_multi --train_for_env_steps=5000000000 --algo=APPO --use_rnn=False --num_workers=36 --num_envs_per_worker=4 --learning_rate=0.0001 --ppo_clip_value=5.0 --recurrence=1 --nonlinearity=tanh --actor_critic_share_weights=False --policy_initialization=xavier_uniform --adaptive_stddev=False --hidden_size=256 --quads_neighbor_hidden_size=256 --quads_neighbor_encoder_type=attention --neighbor_obs_type=pos_vel --with_vtrace=False --max_policy_lag=100000000 --gae_lambda=1.00 --max_grad_norm=5.0 --exploration_loss_coeff=0.0 --rollout=128 --batch_size=1024 --quads_use_numba=True --quads_num_agents=8 --quads_episode_duration=15.0 --quads_mode=mix --quads_formation_size=0.0 --encoder_custom=quad_multi_encoder_deepset --with_pbt=False --quads_collision_reward=5.0 --quads_collision_hitbox_radius=2.0 --quads_collision_falloff_radius=4.0 --quads_collision_smooth_max_penalty=10.0',
```
Comment (alex-petrenko) [inline runner/runs/quad_multi_mixed_vel_col_func.py:12]: Any reason these use the deepsets encoder? Attention wasn't merged yet?

**Review by alex-petrenko — APPROVED** [review]

File: runner/runs/quad_multi_mixed_vel_col_func.py (hunk 2 above)
Comment (Zhehui-Huang) [reply runner/runs/quad_multi_mixed_vel_col_func.py:12]: So, it's not the encoder_custom=quad_multi_encoder_deepset that decides which neighbor encoder that we use.

Comment (Zhehui-Huang) [reply runner/runs/quad_multi_mixed_vel_col_func.py:12]: We use quads_neighbor_encoder_type to decide. And I have already set quads_neighbor_encoder_type=attention here.

Comment (alex-petrenko) [reply runner/runs/quad_multi_mixed_vel_col_func.py:12]: Oh yeah. We should definitely fix this.

Comment (Zhehui-Huang) [reply runner/runs/quad_multi_mixed_vel_col_func.py:12]: How about merge it first? And we can discuss it tomorrow. And I can take charge of it, if you want.

Comment (alex-petrenko) [reply runner/runs/quad_multi_mixed_vel_col_func.py:12]: Which PR? This one I merged.

Comment (Zhehui-Huang) [reply runner/runs/quad_multi_mixed_vel_col_func.py:12]: I mean the current PR, since you have already merged, just ignore my reply.

Comment (alex-petrenko) [reply runner/runs/quad_multi_mixed_vel_col_func.py:12]: You should go rest :) 
I recognise this is a lot to work!! Sleep well and see you tomorrow :)

**Pull request merged**